yt/yt/server/lib/alert_manager/alert_manager.cpp

```diff
@@ -61,56 +76,184 @@ class TAlertManager
             ConvertToYsonString(newConfig, EYsonFormat::Text));
     }
 
-private:
-    const IInvokerPtr ControlInvoker_;
-    const NYTree::IYPathServicePtr OrchidService_;
-
-    TAlertManagerDynamicConfigPtr DynamicConfig_;
-    NConcurrency::TPeriodicExecutorPtr AlertCollectionExecutor_;
-
-    std::vector<TAlert> Alerts_;
-
     void CollectAlerts()
     {
-        VERIFY_SERIALIZED_INVOKER_AFFINITY(ControlInvoker_);
+        std::vector<TAlert> rawAlerts;
+        PopulateAlerts_.Fire(&rawAlerts);
+
+        struct TAggregatedAlert
+        {
+            std::optional<TErrorCode> ErrorCode;
+            std::optional<TString> Description;
+            std::vector<TError> Errors;
+        };
+        THashMap<TString, TAggregatedAlert> categoryToAggregatedAlerts;
 
-        std::vector<TAlert> alerts;
-        PopulateAlerts_.Fire(&alerts);
-        Alerts_.swap(alerts);
+        THashMap<TString, THashSet<TTagList>> uniqueAlerts;
 
-        THashSet<TString> encounteredCategories;
+        for (const auto& rawAlert: rawAlerts) {
+            // Each category + tags combination should be unique.
+            InsertOrCrash(uniqueAlerts[rawAlert.Category], rawAlert.Tags);
 
-        for (const auto& alert : Alerts_) {
-            InsertOrCrash(encounteredCategories, alert.Category);
+            auto& aggregatedAlert = categoryToAggregatedAlerts[rawAlert.Category];
 
-            YT_LOG_WARNING(alert.Error);
+            // Error codes and descriptions should be equal for alerts in the same category.
+            if (!aggregatedAlert.ErrorCode) {
+                aggregatedAlert.ErrorCode = rawAlert.ErrorCode;
+            } else {
+                YT_VERIFY(*aggregatedAlert.ErrorCode == rawAlert.ErrorCode);
+            }
+            // TODO(achulkov2): Avoid duplicating description strings? Doesn't seem to be worth the hassle for now.
```

Comment:
Hmm, but `FormatEnum` requires knowledge about the enum type, which we don't have in `StageAlert`.

Actually, we can probably get rid of `CreateAlert` altogether and make `StageAlert` templated. We don't really want to allow creating alerts in any other way.

Reply:
Eh, I tried doing this and realized that since `StageAlert` is a virtual method it cannot be templated :/
So we really have no way of checking the formatting without over-engineering with type-erasure.

I guess I will have to leave this as is.